module/zfs/arc.c

```diff
@@ -8178,8 +8178,14 @@ l2arc_read_done(zio_t *zio)
 	if (cb->l2rcb_abd != NULL) {
 		ASSERT3U(arc_hdr_size(hdr), <, zio->io_size);
 		if (zio->io_error == 0) {
-			abd_copy(hdr->b_l1hdr.b_pabd, cb->l2rcb_abd,
-			    arc_hdr_size(hdr));
+			if (BP_IS_ENCRYPTED(&cb->l2rcb_bp) &&
+			    (cb->l2rcb_flags & ZIO_FLAG_RAW_ENCRYPT)) {
```

Comment:
This change looks good to me.

I think there's also an opportunity to consolidate how we determine whether we're using b_rabd vs b_pabd.  We figure that out in 3 different places (here, with the same code on line 8205, and then when we calculate using_rdata).  Maybe we should do something like this early in the func, and then just use it:
```
boolean_t using_rdata =
    BP_IS_ENCRYPTED(&cb->l2rcb_bp) && (cb->l2rcb_flags & ZIO_FLAG_RAW_ENCRYPT);
```

Reply:
Fair enough. Fixed.